lib/core/state/mongo-users-state.js

```diff
@@ -0,0 +1,44 @@
+const UserCommands = require('../commands/user-commands')
+const MongoClient = require('../clients/mongodb-client')
+const logger = require('../../helpers/logger')
+
+class MongoUsersState {
+  /**
+   *
+   * @param {Object} settings Settings
+   * @param {String} settings.users users to create
+   */
+  constructor(settings) {
+    const { connection } = settings
+    this.users = settings.users
+    // Client
+    this.client = settings.client || new MongoClient(connection.uri)
+  }
+
+  async sync() {
+    const responses = []
+    // this should sync the users
+    // for now it only creates/updates new ones
+    const userCommands = new UserCommands(this.client)
+
+    for (let i = 0; i < this.users.length; i += 1) {
+      const user = this.users[i]
+      logger.debug('checking if user exits', user.user)
+      const response = await userCommands.usersInfo(user.user)
+      logger.debug('user resp', response)
+      // If this array is filled it means the user exists
+      if (response.users.length) {
+        logger.debug('updating user', user)
+        const resp = await userCommands.updateUser(user)
+        responses.push(resp)
```

Comment:
Hydra should deal with responses, this currently fully swallows errors. See https://github.com/unumotors/mongo-hydra/blob/9ae392709da9f992abd1fb496a0180d6dab14490/lib/core/state/replication.js#L31

Reply:
This doesn't swallow them it returns them. I am happy to switch to exception based handling but  this has ability to show which users were successful.

The problem with exception based flows is we can't automatically disconnect easily...

Reply:
The issue is currently see is that it "looks" like everything is wrong as there will just be a { ok: 0 } in a bunch of code

I'd rather fail hard here and make the user notice super hard that something went wrong?

Reply:
Done